hstream/src/HStream/Server/Core/Stream.hs

```diff
@@ -270,17 +274,31 @@ trimShards ServerContext{..} streamName recordIds = do
   let streamId = S.transToStreamName streamName
   shards <- M.elems <$> S.listStreamPartitions scLDClient streamId
   concurrentCap <- getNumCapabilities
-  res <- limitedMapConcurrently (min 8 concurrentCap) (trim shards) points
-  return $ M.fromList res
+  (errors, res) <- partitionEithers <$> limitedMapConcurrently (min 8 concurrentCap) (trim shards) points
+  if null errors
+    then return $ M.fromList res
+    else throwIO @HE.SomeHStreamException . fromJust . fromException $ head errors
  where
-   trim shards r@Rid{..} = do
-     unless (rShardId `elem` shards) $
-       throwIO . HE.ShardNotExists $ "shard " <> show rShardId <> " doesn't belong to stream " <> show streamName
-     S.trim scLDClient rShardId (rBatchId - 1)
-     Log.info $ "trim to " <> Log.build (show $ rBatchId - 1)
-             <> " for shard " <> Log.build (show rShardId)
-             <> ", stream " <> Log.build streamName
-     return (rShardId, T.pack . show $ r)
+   trim shards r@Rid{..}
+     | rShardId `elem` shards = do
+         try (S.trim scLDClient rShardId (rBatchId - 1)) >>= \case
+           Left (e:: SomeException)
```

Comment:
Why not use `SomeStoreException`?

Reply:
I tried SomeStoreException first, but I didn't find a way to pattern match SomeStoreException to a specific exception, such as S.TOOBIG

Reply:
I haven't looked at the details, but I think at least `catches` can do this job. You do not need to do pattern match.

Reply:
> Does S.trim return exceptions other than SomeHStoreException

Yes, it's possible, but seldom.

> if not then catching SomeException and SomeHStoreException is actually the same

No, they are not the same, for example, while dealing with `AsyncException`

>  so if it returns something else, even if I use catches, I still have to match SomeException

It depends on. But mostly, this is what you should not do.